cfpb/cfgov-refresh
# Promote Expandables from molecule to organism

The Design Ops team has been given blessing by @schaferjh to get Wells inside of Expandables, one way or another, before our Auto Loans project launches next Wednesday. In discussion with @anselmbradford, we decided that the most "atomic" way to do this was to promote Expandables to an organism, so that they could contain other organisms (e.g., Wells). (The other option would be to demote Wells to a molecule, but there are ostensibly plans for Wells to be more complicated in the future.)

This PR makes that happen.

## Additions

- Ability to use Wells inside Expandables 

## Changes

- Everything about Expandable that is needed to convert it from a molecule to an organism.

## Testing

1. Pull branch
2. `./cfgov/manage.py migrate`
3. `gulp styles && gulp scripts`
4. `./cfgov/manage.py runserver`
5. Look at any page that has expandables and verify that they are still correctly functioning.
6. Log in to Wagtail and create a test page (Browse basic) with an Expandable Group, and put a Well inside one of its Expandables.
7. Test that page out, verify everything works and that there are no JS errors.

## Review

- @jimmynotjim 
- @KimberlyMunoz 
- @anselmbradford 
- @sebworks 
- @kurtw 

## Screenshots

<img width="893" alt="screen shot 2016-05-31 at 11 26 09" src="https://cloud.githubusercontent.com/assets/1044670/15680172/9743d9ec-2722-11e6-832d-0d772d9f81e7.png">


## Notes

~~**One major question:**

#2055 set up atomic initialization flags. The init functions were all set to return undefined if they had already been initialized. After I converted Expandables to an organism, their inits started happening _before_ ExpandableGroups, and a JS error was being thrown when ExpandableGroups went to init. The error would prevent the accordion-style groups from only keeping one open at a time.

Because I couldn't figure out how to change the order of the initializations so that ExpandableGroups would keep being initialized first, I "solved" this by, on a hunch, setting the Expandable init function to `return this;` if it found it had already been initialized. This returned the correct Expandable object to the ExpandableGroup init and let it proceed.

If this is kosher, the `return this;` should, I think, be propagated to all of the atomic init functions? If not, I need advice on how to change the order of the initializations so that ExpandableGroups go before Expandables.

Unfortunately, @anselmbradford is out this week, and this PR is pretty time sensitive. @jimmynotjim and @KimberlyMunoz, what do you think?~~

**Update:** Solved with #2171!

## Todos

- Confirm with @duelj and @ajbush whether the padding at the bottom of a well is supposed to be 30px (what the written spec says) or 60px (what the image on the spec page shows)

## Checklist
